sdk/modelsrepository/Azure.Iot.ModelsRepository/src/Fetchers/RemoteModelFetcher.cs

```diff
@@ -40,50 +131,88 @@ public async Task<FetchResult> FetchAsync(string dtmi, Uri repositoryUri, Cancel
 
-            string remoteFetchError = string.Empty;
-            while (work.Count != 0 && !cancellationToken.IsCancellationRequested)
+            return work;
+        }
+
+        private static string GetPath(string dtmi, Uri repositoryUri, bool expanded = false)
+        {
+            string absoluteUri = repositoryUri.AbsoluteUri;
+            return DtmiConventions.DtmiToQualifiedPath(dtmi, absoluteUri, expanded);
+        }
+
+        private string EvaluatePath(string path, CancellationToken cancellationToken = default)
+        {
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope("RemoteModelFetcher.EvaluatePath");
+            scope.Start();
+
+            try
             {
-                string tryContentPath = work.Dequeue();
-                ResolverEventSource.Shared.FetchingModelContent(tryContentPath);
+                using HttpMessage message = CreateGetRequest(path);
 
-                string content = await EvaluatePathAsync(tryContentPath, cancellationToken).ConfigureAwait(false);
-                if (!string.IsNullOrEmpty(content))
+                _pipeline.Send(message, cancellationToken);
+
+                if (message.Response.Status >= 200 && message.Response.Status <= 299)
                 {
-                    return new FetchResult()
-                    {
-                        Definition = content,
-                        Path = tryContentPath
-                    };
+                    return GetContent(message.Response.ContentStream);
+                }
+                else
+                {
+                    throw _clientDiagnostics.CreateRequestFailedException(message.Response);
                 }
-
-                ResolverEventSource.Shared.ErrorFetchingModelContent(tryContentPath);
-                remoteFetchError = string.Format(CultureInfo.CurrentCulture, StandardStrings.ErrorFetchingModelContent, tryContentPath);
             }
-
-            throw new RequestFailedException(remoteFetchError);
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
         }
 
-        private static string GetPath(string dtmi, Uri repositoryUri, bool expanded = false)
+        private async Task<string> EvaluatePathAsync(string path, CancellationToken cancellationToken = default)
         {
-            string absoluteUri = repositoryUri.AbsoluteUri;
-            return DtmiConventions.DtmiToQualifiedPath(dtmi, absoluteUri, expanded);
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope("RemoteModelFetcher.EvaluatePath");
+            scope.Start();
+
+            try
+            {
+                using HttpMessage message = CreateGetRequest(path);
+
+                await _pipeline.SendAsync(message, cancellationToken).ConfigureAwait(false);
+
+                if (message.Response.Status >= 200 && message.Response.Status <= 299)
```

Comment:
Probably should confirm there is a content stream. If the service returned 204, for instance, there wouldn't be a body.

Reply:
Talked offline with Paymaun, he agrees that we should only accept 200's for the GET response. changing the code.